docker/android-system.sh

```diff
@@ -46,64 +79,474 @@ EOF
+
+    purge_packages
+
+    popd
+
+    rm -rf "${td}"
+    rm -rf "${PYTHON_TMPDIR}"
+    rm "${0}"
+}
+
+# java isn't required for the build, but the build expects to
+# find a java compiler. the supported android versions are:
+# https://source.android.com/docs/setup/start/older-versions
+#   Android 7: OpenJDK-8
+fake_java() {
+    local java_type=
+    local java_version=
+    local jre_info=
+    local build_info=
+
+    case "${MAJOR_VERSION}" in
+        5|6)
+            java_type=java
+            java_version=1.7.0
+            jre_info="IcedTea 2.6.9"
+            build_info="build 24.131-b00, mixed mode"
+            ;;
+        *)
+            java_type=openjdk
+            java_version=1.8.0_342
+            jre_info="build 1.8.0_342-8u342-b07-0ubuntu1~20.04-b07"
+            build_info="build 25.342-b07, mixed mode"
+            ;;
+    esac
+
+    cat << EOF > /usr/bin/java
+#!/usr/bin/env bash
+echo "${java_type} version \"${java_version}\""
+echo "OpenJDK Runtime Environment (${jre_info})"
+echo "OpenJDK 64-Bit Server VM (${build_info})"
+EOF
+
+    cat << EOF > /usr/bin/javac
+#!/usr/bin/env bash
+echo "javac ${java_version}"
+EOF
+
+    chmod +x /usr/bin/java
+    chmod +x /usr/bin/javac
+
+    # more faking
+    export ANDROID_JAVA_HOME=/tmp
+    mkdir -p /tmp/lib/
+    touch /tmp/lib/tools.jar
+}
+
+build_android() {
+    if [[ "${ANDROID_SYSTEM_COMPLETE}" != "1" ]]; then
+        export ALLOW_MISSING_DEPENDENCIES=true
+    fi
+
+    set +u
+    # shellcheck disable=SC1091
+    source build/envsetup.sh
+    lunch "aosp_${ARCH}-user"
+    if [[ "${ANDROID_SYSTEM_COMPLETE}" != "1" ]]; then
+        mmma bionic/
+        mmma external/mksh/
+        mmma system/core/toolbox/
+    else
+        mma
+    fi
+    if [[ "${ANDROID_SYSTEM_COMPLETE}" != "1" ]] && [[ "${MAJOR_VERSION}" -ge 11 ]]; then
+        # for some reason, building bionic doesn't build linker64 on the first pass
+        # doing a partial build and a rebuild is just as fast though.
+        rm -rf out/target/product/generic
+        mmma bionic/
+    fi
+    set -u
+}
+
+install_android() {
+    local outdir=
+    if [[ "${ARCH}" = "arm" ]]; then
+        outdir=out/target/product/generic
+    else
+        outdir="out/target/product/generic_${ARCH}"
+    fi
+    mv "${outdir}/system/" /
+    if [[ "${ANDROID_SYSTEM_COMPLETE}" == "1" ]] && [[ -d "${outdir}/apex" ]]; then
+        # can use the APEX linker, no need to use the bootstrap one
+        mv "${outdir}/apex/" /
+    elif [[ "${MAJOR_VERSION}" -ge 10 ]]; then
+        symlink_bootstrap
+    fi
+
+    # list from https://elinux.org/Android_toolbox
+    local tool=
+    for tool in "${@}"; do
+        if [[ ! -f "/system/bin/${tool}" ]]; then
+            ln -s /system/bin/toolbox "/system/bin/${tool}"
+        fi
+    done
+
+    echo "127.0.0.1 localhost" > /system/etc/hosts
+}
+
+symlink_bootstrap() {
+    # for Android 10+, we need to use the bootstrap linker rather than
```

Comment:
All the linkers are now installed in `/system/bin/bootstrap/` for `linker`, `linker64`, `linker_asan`, etc. The ones in `/system/bin/` are broken symlinks to `/apex/com.android.runtime/...`, which means we have to replace these symlinks with the bootstrap ones. This works perfectly fine, however.